design/20190708.certificate-request-crd.md

```diff
@@ -0,0 +1,242 @@
+  - [Graduation Criteria](#graduation-criteria)
+      - [Alpha](#alpha)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+      - [Removing a deprecated flag](#removing-a-deprecated-flag)
+  - [Version Skew Strategy](#version-skew-strategy)
+<!-- /toc -->
+
+## Summary
+
+Currently, certificates issued via cert-manager rely on the `Certificate`
+resource being reconciled by the `Certificate` controller. This resource imposes
+limitations on what issuers are able to honour the `Certificate` resource as
+well as other opinionated implementation details.
+
+This proposal adds a new custom resource `CertificateRequest` that contains an
+x509 certificate signing request, a target issuer, and other metadata about the
+request. Each issuer will have their own `CertificateRequest` controller to
+watch for resources that are referencing them. The `Certificate` controller will
+then rely on creating `CertificateRequest`s to resolve it's own Spec.
+
+## Motivation
+
+Currently the required use of the `Certificate` resource means that users are
+forced to:
+
+- expose the signed certificate's private key to the API server
+- be limited to the finite set of issuers as implemented into the cert-manager
+  project, or, "in-tree"
+- adhere to the `Certificate` controller's opinionated implementation, limiting
+  scope for integrations with other projects
+- rely on developers of the cert-manager project for reviews and approval of new
+  issuers
+
+Due to these issues, cert-manager can be often unsuitable for some use
+cases/integrations or users are unsatisfied with some behaviour. Lack of
+exposure of options that a raw x509 certificate signing request provides can
+also be a source of frustration.
+
+With cert-manger maintainers ensuring that all issuers are always fully
+supported and tested, it becomes difficult for new issuers to become
+accepted. Some developers of new issuers would be happy to maintain these
+issuers themselves however is not possible with all issuers belonging in the
+same code base and repository.
+
+### Goals
+
+- Introduce the `CertificateRequest` resource.
+- Create a `CertificateRequest` controller for each in-tree issuer to resolve
+  `CertificateRequest`.
+- Change the implementation of the `Certificate` controller to rely on the
+  `CertificateRequest` resource to resolve the request.
+- Update documentation detailing this new behaviour and how it can be used to
+  develop out-of-tree implantations of an issuer `CertificateRequest`
+  controller.
+- Create a boilerplate/scaffolding example code to help quick start developers
+  on creating a controller with best practices.
+
+### Non-Goals
+
+- This proposal does not document or explore possible or planned integrations
+  using this new functionality.
+- This proposal will not investigate possible alignment or merging with the
+  Kubernetes internal `CertificateSigningRequest` resource.
```

Comment:
I think we should explicitly note that it *is* something we are interested in, however our motivation is mostly in order to get a built-in approval workflow for CertificateRequests. The feasibility of being able to implement a solution using the built-in type in the near future however is small, so we'd rather 'trail-blaze' here and then try and fold our changes back upstream at a later date.